File: lib/opendmm/search_engines/jav_library.rb
base_uri 'www.javlibrary.com'

def self.search_url(name)
  "/ja/vl_searchbyid.php?keyword=#{CGI::escape(name)}"
end

def self.product_url(name)
  search_page = get_with_retry search_url(name)
  return nil unless search_page
  if search_page.code == 302
    url = search_page.headers['location']
  else
    search_html = Utils.html_in_utf8 search_page
    first_result = search_html.at_css('#rightcolumn > div.videothumblist > div.videos > div.video > a')
    return nil unless first_result
    url = first_result['href']
  end
  URI.join(search_page.request.last_uri, url).to_s
end

def self.parse_product_html(html)
  {
    actresses:       html.css('#video_cast .text span.cast').map(&:text),
    code:            html.css('#video_id .text').text,
    cover_image:     html.at_css('#video_jacket > a')['href'],
    directors:       html.css('#video_director .text span.director').map(&:text),
    genres:          html.css('#video_genres .text span.genre').map(&:text),
    label:           html.css('#video_label .text').text,
    maker:           html.css('#video_maker .text').text,
    movie_length:    html.css('#video_length .text').text + ' minutes',
    release_date:    html.css('#video_date .text').text,
    thumbnail_image: html.at_css('#video_jacket_img')['src'],
    title:           html.css('#video_title > h3').text.remove(html.css('#video_id .text').text),
  }
end

File: javlibrary.go
package opendmm

import (
	"fmt"
	"net/http"
	"net/url"
	"sync"
	"time"

	"github.com/junzh0u/httpx"

	"github.com/PuerkitoBio/goquery"
	"github.com/golang/glog"
)

var javCookie = http.Cookie{
	Name:     "over18",
	Value:    "18",
	Domain:   "www.javlibrary.com",
	Path:     "/",
	HttpOnly: false,
	Secure:   false,
	Expires:  time.Now().Add(1000 * time.Hour),
}

func javSearch(query string, wg *sync.WaitGroup, metach chan MovieMeta) {
	keywords := dmmGuess(query)
	for keyword := range keywords.Iter() {
		wg.Add(1)
		go func(keyword string) {
			defer wg.Done()
			javSearchKeyword(keyword, wg, metach)
		}(keyword.(string))
	}
}

func javSearchKeyword(keyword string, wg *sync.WaitGroup, metach chan MovieMeta) {
	glog.Info("Keyword: ", keyword)
	urlstr := fmt.Sprintf(
		"http://www.javlibrary.com/ja/vl_searchbyid.php?keyword=%s",
		url.QueryEscape(keyword),
	)
	javParse(urlstr, keyword, wg, metach)
}

func javParse(urlstr string, keyword string, wg *sync.WaitGroup, metach chan MovieMeta) {
	glog.V(2).Info("Product/Search page: ", urlstr)
	doc, err := newDocument(urlstr, httpx.GetContentViaPhantomJS([]*http.Cookie{&javCookie}, 6*time.Second, "Start of Content", "Checking your browser before accessing"))
	if err != nil {
		glog.V(2).Infof("Error parsing %s: %v", urlstr, err)
		return
	}

	var meta MovieMeta
	var ok bool
	meta.Page, ok = doc.Find("link[rel=shortlink]").Attr("href")
	if ok {
		meta.Code = doc.Find("#video_id .text").Text()
		meta.Title = doc.Find("#video_title > h3").Text()
		meta.CoverImage, _ = doc.Find("#video_jacket > img").Attr("src")
		meta.ReleaseDate = doc.Find("#video_date .text").Text()
		meta.MovieLength = doc.Find("#video_length .text").Text()
		meta.Directors = doc.Find("#video_director .text span.director").Map(
			func(i int, span *goquery.Selection) string {
				return span.Text()
			})
		meta.Maker = doc.Find("#video_maker .text").Text()
		meta.Label = doc.Find("#video_label .text").Text()
		meta.Genres = doc.Find("#video_genres .text span.genre").Map(
			func(i int, span *goquery.Selection) string {
				return span.Text()
			})
		meta.Actresses = doc.Find("#video_cast .text span.cast span.star").Map(
			func(i int, span *goquery.Selection) string {
				return span.Text()
			})

		if !dmmIsCodeEqual(keyword, meta.Code) {
			glog.V(2).Infof("Code mismatch: Expected %s, got %s", keyword, meta.Code)
		} else {
			metach <- meta
		}
	} else {
		urlbase, err := url.Parse(urlstr)
		if err != nil {
			glog.V(2).Info(err)
			return
		}
		doc.Find("div.videothumblist > div.videos > div.video > a").Each(
			func(i int, a *goquery.Selection) {
				href, ok := a.Attr("href")
				if !ok {
					return
				}
				urlhref, err := urlbase.Parse(href)
				if err != nil {
					glog.V(2).Info(err)
					return
				}
				wg.Add(1)
				go func() {
					defer wg.Done()
					javParse(urlhref.String(), keyword, wg, metach)
				}()
			})
	}
}
